File: C++/patternMatch.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
#include <ctime>
#include <set>
#include <utility>
#include <algorithm>
#include <map>

using namespace std;

bool getNewPatternMap(map<char, string> &m, map<char, string> &res, char c, string &s) {
  if (m.find(c) != m.end()) {
    if (m[c] != s) {
      return false;
    }
  } else {
    for (auto it : m) {
      if (it.second == s && it.first != c) {
        return false;
      }
    }
  }
  res = m;
  res[c] = s;
  return true;
}

bool isPatternMatched(string& str, string& pattern) {
  if (pattern.size() < 1) {
    return true;
  }
  vector<vector<map<char, string> > > res1(str.size(), vector<map<char, string> >()), res2(str.size(), vector<map<char, string> >());
  for (int i = 0; i < res1.size(); i++) {
    map<char, string> m;
    m[pattern[0]] = str.substr(0, i + 1);
    res1[i].push_back(m);
  }

  for (int i = 1; i < pattern.size(); i++) {
    for (int j = 0; j < str.size(); j++) {
      res2[j].clear();
      for (int k = 1; k <= j; k++) {
        string lastWord = str.substr(k, j - k + 1);
        for (map<char, string> m : res1[k - 1]) {
          map<char, string> tmp;
          if (getNewPatternMap(m, tmp, pattern[i], lastWord)) {
            res2[j].push_back(tmp);
          }
        }
      }
    }
    swap(res1, res2);
  }
  if (res1[str.size() - 1].size() > 0) {
    for (map<char, string> m : res1[str.size() - 1]) {
      for (auto p : m) {
        cout << p.first << ":" << p.second << endl;
      }
      cout << endl;
    }
  }
  return res1[str.size() - 1].size() > 0;
}

int main(void) {
  string str = "asdasdasdasd", pattern = "AAAA";
  cout << isPatternMatched(str, pattern) << endl;
  str = "bigboyboybig";
  pattern = "ABBA";
  cout << isPatternMatched(str, pattern) << endl;
  str = "bigbigbigbig";
  cout << isPatternMatched(str, pattern) << endl;
}

File: java/test/src/problems/PatternMatch.java
package problems;

import java.util.ArrayList;
import java.util.HashMap;

public class PatternMatch {
    public boolean isPatternMatched(String str, String pattern) {
        if (pattern.length() < 1) {
            return true;
        }
        ArrayList<ArrayList<HashMap<Character, String>>> res1 = new ArrayList<ArrayList<HashMap<Character, String>>>(str.length());
        ArrayList<ArrayList<HashMap<Character, String>>> res2 = new ArrayList<ArrayList<HashMap<Character, String>>>(str.length());
        for (int i = 0; i < str.length(); i++) {
            res1.add(i, new ArrayList<HashMap<Character, String>>());
            res2.add(i, new ArrayList<HashMap<Character, String>>());
            HashMap<Character, String> m = new HashMap<Character, String>();
            m.put(pattern.charAt(0), str.substring(0, i + 1));
            res1.get(i).add(m);
        }
        for (int i = 1; i < pattern.length(); i++) {
            for (int j = 0; j < str.length(); j++) {     
                res2.get(j).clear();
                for (int k = 1; k <= j; k++) {
                    String lastWord = str.substring(k, j + 1);
                    for (HashMap<Character, String> m : res1.get(k - 1)) {
                        if (!m.containsKey(pattern.charAt(i))) {
                            HashMap<Character, String> m1 = new HashMap<Character, String>(m);
                            m1.put(pattern.charAt(i), lastWord);
                            res2.get(j).add(m1);
                        } else if (m.get(pattern.charAt(i)).equals(lastWord)) {
                            HashMap<Character, String> m1 = new HashMap<Character, String>(m);
                            res2.get(j).add(m1);
                        }
                    }
                }
            }
            ArrayList<ArrayList<HashMap<Character, String>>> res3 = res1;
            res1 = res2;
            res2 = res3;
        }
        for (HashMap<Character, String> m: res1.get(str.length() - 1)) {
            for (Character key : m.keySet()) {
                System.out.println(key + ":" + m.get(key));
            }
        }
        return res1.get(str.length() - 1).size() > 0;
    }
    public static void main(String[] args) {
        String str= new String("asdasdasdasd"), pattern = new String("AAAA");
        PatternMatch matcher = new PatternMatch();
        System.out.println(matcher.isPatternMatched(str, pattern));
        str = "bigboyboybig";
        pattern = "ABBA";
        System.out.println(matcher.isPatternMatched(str, pattern));
    }

}
